components/pd_client/src/util.rs

```diff
@@ -180,56 +230,56 @@ impl LeaderClient {
             let start = Instant::now();
             let connector = PdConnector::new(inner.env.clone(), inner.security_mgr.clone());
             let members = inner.members.clone();
-            let fut = async move { connector.reconnect_leader(&members, force).await };
-            slow_log!(start.elapsed(), "PD client try connect leader");
+            let fut = async move {
+                connector
+                    .reconnect_pd(
+                        members,
+                        self.get_forwarded_host(),
+                        force,
+                        self.enable_forwarding,
+                    )
+                    .await
+            };
+            slow_log!(start.elapsed(), "try reconnect pd");
             (fut, start)
         };
 
-        let (client, members) = match future.await? {
-            Some(pair) => pair,
+        let (client, forwarded_host, members) = match future.await? {
+            Some(tuple) => tuple,
             None => return Ok(()),
         };
-        fail_point!("leader_client_reconnect", |_| Ok(()));
+        fail_point!("pd_client_reconnect", |_| Ok(()));
 
-        {
-            let start_refresh = Instant::now();
-            let mut inner = self.inner.wl();
-            let (tx, rx) = client.region_heartbeat().unwrap_or_else(|e| {
-                panic!("fail to request PD {} err {:?}", "region_heartbeat", e)
-            });
-            info!("heartbeat sender and receiver are stale, refreshing ...");
-
-            // Try to cancel an unused heartbeat sender.
-            if let Either::Left(Some(ref mut r)) = inner.hb_sender {
-                debug!("cancel region heartbeat sender");
-                r.cancel();
-            }
-            inner.hb_sender = Either::Left(Some(tx));
-            let prev_receiver = std::mem::replace(&mut inner.hb_receiver, Either::Left(Some(rx)));
-            let _ = prev_receiver.right().map(|t| t.wake());
-            inner.client_stub = client;
-            inner.members = members;
-            inner.last_update = Instant::now();
-            if let Some(ref on_reconnect) = inner.on_reconnect {
-                on_reconnect();
-            }
-            slow_log!(
-                start_refresh.elapsed(),
-                "PD client refresh region heartbeat",
-            );
-        }
-        info!("updating PD client done"; "spend" => ?start.elapsed());
+        self.update_client(client, forwarded_host, members);
+        info!("tring to update PD client done"; "spend" => ?start.elapsed());
         Ok(())
     }
 }
 
+// The call option here doesn't need the timeout setting which is used by stream.
+fn call_option(forwarded_host: &str) -> CallOption {
```

Comment:
The name is not descriptive. How about `pd_forward_call_option`?

Reply:
IMO, this option is not only for forwarding, and since it is in pd client, we don't need to emphasize it is the call option for pd.